velox/functions/prestosql/window/Ntile.cpp

```diff
@@ -0,0 +1,225 @@
+    // a bucket as the (number of rows in partition) / (number of buckets).
+    int64_t rowsPerBucket;
+    // There could be some buckets with rowsPerBucket + 1 number of rows,
+    // as the partition rows might not be exactly divisible
+    // by the number of buckets. There are
+    // (number of rows in partition) % (number of buckets) such buckets.
+    int64_t bucketsWithExtraRow;
+    // When assigning bucket numbers, the first 'bucketsWithExtraRow' buckets
+    // will have (rowsPerBucket + 1) rows. This row number at this boundary is
+    // extraBucketsBoundary = bucketsWithExtraRow * (rowsPerBucket + 1). Beyond
+    // this row number in the partition, the buckets will have only
+    // rowsPerBucket number of rows. This boundary is useful when computing the
+    // bucket value.
+    int64_t extraBucketsBoundary;
+  };
+
+  inline const BucketMetrics computeBucketMetrics(int64_t numBuckets) {
+    auto rowsPerBucket = numPartitionRows_ / numBuckets;
+    auto bucketsWithExtraRow = numPartitionRows_ % numBuckets;
+    auto extraBucketsBoundary = (rowsPerBucket + 1) * bucketsWithExtraRow;
+    return {rowsPerBucket, bucketsWithExtraRow, extraBucketsBoundary};
+  }
+
+  inline int64_t computeBucketValue(
+      vector_size_t rowNumber,
+      const BucketMetrics& bucketMetrics) {
+    if (rowNumber < bucketMetrics.extraBucketsBoundary) {
+      return rowNumber / (bucketMetrics.rowsPerBucket + 1) + 1;
+    }
+    return (rowNumber - bucketMetrics.bucketsWithExtraRow) /
+        bucketMetrics.rowsPerBucket +
+        1;
+  }
+
+  void computeNtileFromColumn(
+      vector_size_t numRows,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    bucketVector_->resize(numRows);
+    partition_->extractColumn(
+        bucketColumn_.value(), partitionOffset_, numRows, 0, bucketVector_);
+
+    auto* resultFlatVector = result->asFlatVector<int64_t>();
+    auto* rawValues = resultFlatVector->mutableRawValues();
+    for (auto i = 0; i < numRows; i++) {
+      vector_size_t rowIndex = i + partitionOffset_;
+      if (bucketFlatVector_->isNullAt(i)) {
+        resultFlatVector->setNull(resultOffset + i, true);
+      } else {
+        auto numBuckets = bucketFlatVector_->valueAt(i);
+        VELOX_USER_CHECK(numBuckets >= 1, kBucketErrorString);
+        auto bucketsMoreThanPartition = numBuckets > numPartitionRows_;
+        if (bucketsMoreThanPartition) {
+          rawValues[resultOffset + i] = rowIndex + 1;
+        } else {
+          rawValues[resultOffset + i] =
+              computeBucketValue(rowIndex, computeBucketMetrics(numBuckets));
+        }
+      }
+    }
+  }
+
+  void computeNtileWithConstants(
+      vector_size_t numRows,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    if (numFixedBuckets_.has_value()) {
+      auto* rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
+      if (fixedBucketsMoreThanPartition_) {
+        for (auto i = 0; i < numRows; i++) {
```

Comment:
This loop is inefficient. Use std::iota.

Reply:
I didn't entirely follow this suggestion. Why is itoa needed here ?

Reply:
std::iota: "Fills the range [first, last) with sequentially increasing values, starting with value and repetitively evaluating ++value."

My reading is that this loop fills rawValues buffer with sequentially increasing values, which is what std::iota does.

Reply:
My bad...I read this is itoa (and not iota). Changed to use iota.